tokio/src/task/scope.rs

```diff
@@ -0,0 +1,439 @@
+    pub fn child(&self) -> Scope {
+        Scope::with_parent(self.clone())
+    }
+
+    /// Spawns a task on the scope
+    pub fn spawn<'inner, T, R>(&'inner self, task: T) -> ScopedJoinHandle<'inner, R>
+    where
+        T: Future<Output = R> + Send + 'static,
+        R: Send + 'static,
+        T: 'inner,
+    {
+        let spawn_handle =
+            crate::runtime::context::spawn_handle().expect("Spawn handle must be available");
+
+        let releaser = self.scope.wait_group.add();
+
+        let child_task = {
+            spawn_handle.spawn(async move {
+                // Drop this at the end of the task to signal we are done and unblock
+                // the WaitGroup
+                let _wait_group_releaser = releaser;
+
+                // Execute the child task
+                task.await
+            })
+        };
+
+        // Since `Scope` is `Sync` and `Send`, cancellations can happen at any time
+        // in case of invalid use. Therefore we only check cancellations once:
+        // After the task has been spawned. Since the cancellation is already set,
+        // we need to wait for the task to complete. Then we panic due to invalid
+        // API usage.
+        if self.scope.is_cancelled() {
+            futures::executor::block_on(async {
+                let _ = child_task.await;
+            });
+            panic!("Spawn on cancelled Scope");
+        }
+
+        ScopedJoinHandle {
+            handle: child_task,
+            phantom: core::marker::PhantomData,
+        }
+    }
+
+    /// Spawns a task on the scope, which will get automatically cancelled if
+    /// the `CancellationToken` which is associated with the current `ScopeHandle`
+    /// gets cancelled.
+    pub fn spawn_cancellable<'inner, T, R>(
+        &'inner self,
+        task: T,
+    ) -> ScopedJoinHandle<'inner, Result<R, CancellationError>>
+    where
+        T: Future<Output = R> + Send + 'static,
+        R: Send + 'static,
+        T: 'inner,
+    {
+        let cancel_token = self.cancellation_token().clone();
+
+        self.spawn(async move {
+            crate::pin!(task);
+            use futures::FutureExt;
+
+            futures::select! {
+                _ = cancel_token.cancelled().fuse() => {
+                    // The child task was cancelled
+                    Err(CancellationError{})
+                },
+                res = task.fuse() => {
+                    Ok(res)
+                }
+            }
+        })
+    }
+}
+
+/// Defines how a scope will behave if the `Future` it returns get dropped
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum ScopeDropBehavior {
+    /// When a scope is dropped while tasks are outstanding, the current thread
+    /// will panic. Since this will not wait for child tasks to complete, the
+    /// child tasks can outlive the parent in this case.
+    Panic,
+    /// When a scope is dropped while tasks are outstanding, the process will be
+    /// aborted.
+    Abort,
```

Comment:
Is there value in having the option vs someone just setting `panic=abort` in their `Cargo.toml`?

Reply:
Not too much. Maybe if someone mainly cares about not misusing `scope`s. It's also a relict from the idea if to be able to borrow from the parent task inside child tasks. In order to guarantee that property a `panic!` wouldn't be strong enough.

That said those APIs are probably all not needed for a v1, and a reasonable default could be picked. The options are mostly presented tho show how a scope could be configured in case configurability is desired.